app/controllers/tweets_controller.rb

```diff
@@ -0,0 +1,86 @@
+class TweetsController < ApplicationController
+
+  get '/tweets' do
+    if logged_in?
+      @tweets = Tweet.all
+      @username = current_user.username
+      erb :"tweets/tweets"
+    else
+      flash[:notice] = "You must be logged in to view this page."
+      redirect to "/login"
+    end
+  end
+
+  get "/tweets/new" do
+    if logged_in?
+      erb :"tweets/create_tweet"
+    else
+      flash[:notice] = "You must be logged in to create a tweet."
+      redirect to "/login"
+    end
+  end
+
+  get "/tweets/:id/edit" do
+    if !logged_in?
+      flash[:notice] = "You must be logged in to edit a tweet."
+      redirect to "/login"  
+    end
+    @tweet = Tweet.find(params[:id])
+    if @tweet && (@tweet.user_id == current_user.id)
+      erb :"tweets/edit_tweet"
+    else
+      flash[:notice] = "You do not have permission to edit tweets you did not create."
+      redirect to "/login"
+    end
+  end
+
+  post "/tweets" do
+    if params["content"].empty?
+      flash[:notice] = "You did not enter any thoughts to create a tweet."
+      redirect to "/tweets/new" 
+    end
+    @tweet = Tweet.new(content: params["content"], user_id:current_user.id)
+    @tweet.save
+    redirect to "/tweets/#{@tweet.id}"
+  end
+
+  get "/tweets/:id" do
+    if logged_in?
+      @tweet = Tweet.find(params[:id])
+      @user = @tweet.user
+      erb :"tweets/show_tweet"
+    else
+      flash[:notice] = "You must be logged in to view this page."
+      redirect to "/login"
+    end
+  end
+
+  patch "/tweets/:id" do
+    if params["content"].empty?
+      flash[:notice] = "You did not enter any thoughts when editing the tweet."
+      redirect to "/tweets/#{params[:id]}/edit" 
+    end
+    @tweet = Tweet.find(params[:id])
+    if @tweet && (@tweet.user_id == current_user.id)
+      @tweet.update(content: params["content"])
+      @tweet.save
+      redirect to "/tweets/#{@tweet.id}"
+    end
+  end
+
+  delete "/tweets/:id/delete" do
+    if !logged_in?
+      flash[:notice] = "You must be logged in to delete a tweet."
+      redirect to "/login"  
+    end
+    @tweet = Tweet.find(params[:id])
+    if @tweet && (@tweet.user_id == current_user.id)
+      @tweet.delete
```

Comment:
Prefer to find tweets that are already only associated to the `current_user` than to manually check the `user_id` foreign key of the tweet against the `current_user.id`

```ruby
tweet = current_user.tweets.find_by(id: params[:id])
if tweet && tweet.destroy
  redirect "/tweets"
else
  redirect "/tweets/#{tweet.id}"
end
```

We're using the association collection method provided by [`has_many`](http://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-has_many) to query the user's tweets. 

Note the usage of [`find_by`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find-by) vs [`find`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find). Had we used `find` here, ActiveRecord would raise an error in the event that we could not find a tweet associated with this user. `find_by` will allow this action to fail gracefully, simply returning nil if the user is trying to delete a tweet that they do not have access to.